### PR TITLE
Add ADR for moving the backend checks to a step function.

### PR DESCRIPTION
**Pull request by MancunianSam**

### CODE REVIEW

File: architecture-decision-records/0027-step-functions-for-backend-checks.md [hunk 1]
```diff
@@ -0,0 +1,42 @@
+# 26. Step functions for backend checks
+
+**Date:** 2023-01-18
+
+## Context
+Once the files in TDR are uploaded, we run antivirus, a server side checksum and a Droid file format check.
+
+We also need to set various file and consignment statuses based on the results of the checks.
+
+We also need to now the status of any redacted files in the consignment.
```
Comment (TomJKing) [inline architecture-decision-records/0027-step-functions-for-backend-checks.md:10]: Spelling: `know` not `now`

File: architecture-decision-records/0027-step-functions-for-backend-checks.md [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+* It is difficult to check statuses which apply across all files in a consignment, for example, all files need to be checked for redacted status.
+* The API is called once for each file. This causes a lot of load on the database. 
+
+### Use a step function
+The step function will run once per consignment. 
+It will be triggered by API Gateway and authenticated with the existing consignment export authoriser lambda.
+
+#### Advantages
+* It is run once per consignment which won't block smaller transfers with larger ones. 
+* We can run a lambda once all the file checks are complete which means updating consignment statuses which apply to all files easier.
```
Comment (TomJKing) [inline architecture-decision-records/0027-step-functions-for-backend-checks.md:32]: Maybe reword slightly: `We can run a lambda once all the file checks are complete, this makes it easier to update consignment statuses, which require to all files statuses.`

File: architecture-decision-records/0027-step-functions-for-backend-checks.md [hunk 3]
```diff
@@ -0,0 +1,42 @@
...
+* It allows a lot of the status updates related to the backend checks to be decoupled from the API.
+* It allows us to centralise the error handling, sending errors to the notifications service when there is any exception in the process.
+
+#### Disadvantes
+* It will be more expensive, as we have to pay for step function transitions as well as the lambda invocations we were already paying for.
+
+## Decision
+We have decided to move the backend checks to use step functions. The main reason is that this will prevent larger consignments blocking smaller ones.
```
Comment (TomJKing) [inline architecture-decision-records/0027-step-functions-for-backend-checks.md:40]: Maybe add a something about why this is important: `TDR now supports the transfers of judgments, which means a lot of small, single file transfers happen daily, which was not considered as something TDR would need to support in the original design. In addition there are SLAs regarding the time taken to process judgment transfers`

File: architecture-decision-records/0027-step-functions-for-backend-checks.md [hunk 4]
```diff
@@ -0,0 +1,42 @@
...
+* It allows a lot of the status updates related to the backend checks to be decoupled from the API.
+* It allows us to centralise the error handling, sending errors to the notifications service when there is any exception in the process.
+
+#### Disadvantes
+* It will be more expensive, as we have to pay for step function transitions as well as the lambda invocations we were already paying for.
+
+## Decision
+We have decided to move the backend checks to use step functions. The main reason is that this will prevent larger consignments blocking smaller ones. 
+By moving a lot of the status logic out of the API and into the step function, we can make the API less complicated and easier to maintain.
+As we are only updating the API once per consignmnent, we can run more of the checks in parallel which shoudl reduce the time taken for the checks on large numbers of files.
```
Comment (TomJKing) [inline architecture-decision-records/0027-step-functions-for-backend-checks.md:42]: Spelling: `should`

File: architecture-decision-records/0027-step-functions-for-backend-checks.md [hunk 5]
```diff
@@ -0,0 +1,42 @@
...
+#### Advantages
+* It is run once per consignment which won't block smaller transfers with larger ones. 
+* We can run a lambda once all the file checks are complete which means updating consignment statuses which apply to all files easier.
+* It allows a lot of the status updates related to the backend checks to be decoupled from the API.
+* It allows us to centralise the error handling, sending errors to the notifications service when there is any exception in the process.
+
+#### Disadvantes
+* It will be more expensive, as we have to pay for step function transitions as well as the lambda invocations we were already paying for.
+
+## Decision
```
Comment (TomJKing) [inline architecture-decision-records/0027-step-functions-for-backend-checks.md:39]: Maybe add a point about the cost, and that something would need to be done to prevent the blocking of small transfers, anyway, which would likely mean an increase in cost somewhere, eg better DB instance etc.

**Review by TomJKing — COMMENTED** [review]
Looks good. Thanks for writing this up. Just some minor comments.

Also the diagrams will need updating to reflect the new architecture.

File: architecture-decision-records/0027-step-functions-for-backend-checks.md (hunk 2 above)
Comment (MancunianSam) [reply architecture-decision-records/0027-step-functions-for-backend-checks.md:32]: Did you mean `require to all file statuses` or should that be `require all file statuses`? I've changed it to the second one for now.

File: architecture-decision-records/0027-step-functions-for-backend-checks.md (hunk 3 above)
Comment (MancunianSam) [reply architecture-decision-records/0027-step-functions-for-backend-checks.md:40]: Good point, I've added it in.

File: architecture-decision-records/0027-step-functions-for-backend-checks.md (hunk 5 above)
Comment (MancunianSam) [reply architecture-decision-records/0027-step-functions-for-backend-checks.md:39]: I've put something in about it.

**Comment by MancunianSam** [comment]
I've updated the architecture diagram as well.

**Review by TomJKing — APPROVED** [review]
Looks good. Thanks
